src/dtls/network.c

```diff
@@ -0,0 +1,962 @@
+        return -1;
+    }
+
+    memcpy_s(ubuf.base, len, buf, len);
+    ubuf.len = len;
+    req->data = ubuf.base;
+
+    int err = uv_udp_send(req, cn->socket, &ubuf, 1, (const struct sockaddr *)&cn->peer->addr.inaddr, OnTLSSendComplete);
+    if (err) {
+        DPS_ERRPRINT("ERROR: couldn't send UDP packet: %s\n", uv_err_name(err));
+        return -1;
+    }
+
+    return len;
+}
+
+static void CancelPendingWrites(DPS_NetConnection* cn, PendingWrite* pw)
+{
+    while (pw) {
+        PendingWrite* next = pw->next;
+        pw->sendCompleteCB(cn->node, pw->appCtx, cn->peer, pw->bufs, pw->numBufs, DPS_ERR_NETWORK);
+        free(pw->bufs);
+        free(pw);
+        DPS_DBGPRINT("freeing PendingWrite %p", pw);
+        pw = next;
+    }
+}
+
+static void DestroyConnection(DPS_NetConnection* cn)
+{
+    if (cn->type == MBEDTLS_SSL_IS_SERVER) {
+        mbedtls_ssl_cookie_free(&cn->cookieCtx);
+        mbedtls_ssl_cache_free(&cn->cacheCtx);
+    }
+
+    mbedtls_entropy_free(&cn->entropy);
+    mbedtls_ctr_drbg_free(&cn->randgen);
+    mbedtls_ssl_config_free(&cn->conf);
+    mbedtls_ssl_free(&cn->ssl);
+
+    PendingRead* pr = cn->readQueue;
+    while (pr) {
+        PendingRead* next = pr->next;
+        free(pr);
+        pr = next;
+    }
+
+    CancelPendingWrites(cn, cn->writeQueue);
+    CancelPendingWrites(cn, cn->callbackQueue);
+
+    free(cn);
+}
+
+static DPS_NetConnection* CreateConnection(DPS_Node* node, const struct sockaddr* addr, int type)
+{
+    assert(IsNodeThread(node));
+
+    int ret;
+    DPS_NetConnection* cn;
+
+    if (type == MBEDTLS_SSL_IS_SERVER) {
+        DPS_DBGPRINT("creating a server DTLS context\n");
+    } else {
+        DPS_DBGPRINT("creating a client DTLS context\n");
+    }
+
+    // TODO: Add support for PKI instead (or in addition to) the network key.
+    static const unsigned char id[] = "dps";
+    static const size_t idLen = sizeof(id);
+    uint8_t key[256] = {};
+    size_t keyLen = 0;
+
+    DPS_KeyStore* keyStore = node->keyStore;
+    if (!keyStore || !keyStore->networkKeyCB) {
+        return NULL;
+    }
+    ret = keyStore->networkKeyCB(keyStore, key, sizeof(key), &keyLen);
+    if (ret != DPS_OK) {
+        return NULL;
+    }
+    if (keyLen == 0) {
+        return NULL;
+    }
+
+    cn = calloc(1, sizeof(DPS_NetConnection));
+    if (!cn) {
+        return NULL;
+    }
+
+    cn->node = node;
+    cn->type = type;
+
+    uv_timer_init(node->loop, &cn->timer);
+    cn->timerStatus = -1;
+    cn->timer.data = cn;
+
+    uv_idle_init(DPS_GetLoop(node), &cn->idleForSendCallbacks);
+    cn->idleForSendCallbacks.data = cn;
+
+    if (addr->sa_family == AF_INET) {
+        cn->socket = &node->netCtx->tx4Socket;
+    } else {
+        cn->socket = &node->netCtx->rxSocket;
+    }
+
+    mbedtls_ssl_init(&cn->ssl);
+    mbedtls_ssl_config_init(&cn->conf);
+    mbedtls_ctr_drbg_init(&cn->randgen);
+
+    // TODO: Verify we have good entropy sources.
+    mbedtls_entropy_init(&cn->entropy);
```

Comment:
Perhaps you can leverage the code in uuid.c which has code to initialize entropy using platform-specific calls.